Simplify fab has access lookup

jhtimmins opened the pull request:
Simplifies the FAB security manager code by doing the following.
1. Modifies permission lookup by accessing `User.perms` property method and checking if a permission pair is included in the user's list of perms.
2. Extends `AnonymousUserMixin` to create `AnonymousUser`, which has both `perms` and `roles` property methods. Removes the need for custom auth logic for anonymous users.
3. Removes helper methods that provide little value, including:
 * `.get_readable_dags()`
 * `.current_user_has_permissions()`
 * `._has_resource_access()`
 * `.user_has_perms()`
 * `.user_has_roles()`
4. Removes unused methods:
 * `SecurityManager.is_item_public()`
5. Moves the code to create a root dag resource name to `Permissions.resource_name_for_dag()`

This is the next piece of the permissions refactor.

jhtimmins committed:
Use FAB models.

jhtimmins committed:
Use FAB models.

jhtimmins committed:
Remove incorrect conversions to new permission naming scheme.

jhtimmins committed:
Remove incorrect conversions to new permission naming scheme.

jhtimmins committed:
Fix missing FAB renames.

jhtimmins committed:
Fix missing FAB renames.

jhtimmins committed:
Remove unused FAB compatibility fixes in models.py.

jhtimmins committed:
Remove unused FAB compatibility fixes in models.py.

jhtimmins committed:
Set perms directly on user objects.

jhtimmins committed:
Set perms directly on user objects.

jhtimmins committed:
Set perms properties on User model.

jhtimmins committed:
Set perms properties on User model.

jhtimmins committed:
Rename missed old naming scheme conversion.

jhtimmins committed:
Rename missed old naming scheme conversion.

jhtimmins committed:
Remove unused imports.

jhtimmins committed:
Remove unused imports.

jhtimmins committed:
Remove unused imports.

jhtimmins committed:
Remove unused imports.

jhtimmins committed:
Remeve get_user_roles() method.

jhtimmins committed:
Remeve get_user_roles() method.

andrewgodwin commented on airflow/www/fab_security/sqla/models.py:
```diff
@@ -178,6 +180,13 @@ def is_active(self):
+    def perms(self):
+        perms = set()
+        for role in self.roles:
+            perms.update((perm.action.name, perm.resource.name) for perm in role.permissions)
```
I think you may need to do this as a single query for efficiency reasons?

jhtimmins replied:
I added `lazy=joined` to several model fields, which causes prefetching. It now runs 11 queries instead of 49.

jhtimmins replied:
@andrewgodwin that said, this PR does add some overall time to the tests. `pytest tests/views --with-db-init` runs in 1:53 on main and 2:00 on my branch. How significant is that?

jhtimmins committed:
Make permissions eagerload.

jhtimmins committed:
Make permissions eagerload.

jhtimmins committed:
Remove unused imports.

jhtimmins committed:
Remove unused imports.

uranusjr commented on airflow/api_connexion/endpoints/dag_run_endpoint.py:
```diff
@@ -211,7 +212,7 @@ def get_dag_runs_batch(session):
+    readable_dag_ids = {dag.dag_id for dag in appbuilder.sm.get_readable_dags(g.user)}
     query = session.query(DagRun)
     if data.get("dag_ids"):
         dag_ids = set(data["dag_ids"]) & set(readable_dag_ids)
```
```suggestion
        dag_ids = set(data["dag_ids"]) & readable_dag_ids
```

Because `readable_dag_ids` is now already a set

Or

```suggestion
        dag_ids = readable_dag_ids.intersection(data["dag_ids"])
```

uranusjr commented on airflow/www/fab_security/manager.py:
```diff
@@ -85,6 +85,33 @@ def _oauth_tokengetter(token=None):
+    @property
+    def perms(self):
+        if not self._perms:
+            self._perms = set()
+            for role in self.roles:
+                self._perms.update({(perm.action.name, perm.resource.name) for perm in role.permissions})
```
```suggestion
            self._perms = {
                (perm.action.name, perm.resource.name)
                for role in self.roles
                for perm in role.permissions
            }
```

Same for `User` below

jhtimmins replied:
This is a fair critique, but I have a pretty strong aversion to nested comprehensions, as I'm convinced Python's design is flawed.

jhtimmins committed:
Clarify query params.

jhtimmins committed:
Clarify query params.

jhtimmins committed:
Modify sort logic so MSSQL passes.

jhtimmins committed:
Modify sort logic so MSSQL passes.

jhtimmins committed:
Add text modifier to order_by values.

jhtimmins committed:
Add text modifier to order_by values.

uranusjr commented on airflow/www/auth.py:
```diff
@@ -35,7 +35,7 @@ def decorated(*args, **kwargs):
 
             appbuilder = current_app.appbuilder
-            if not g.user.is_anonymous and not appbuilder.sm.current_user_has_permissions():
+            if not (g.user.is_anonymous or g.user.perms):
```
I feel the previous version is a bit easier to understand (i.e. “not anonymous but does not have any permissions”)

```suggestion
            if not g.user.is_anonymous and not g.user.perms:
```

ashb replied:
See this one @jhtimmins

jhtimmins committed:
Remove calls to get_*_dags.

jhtimmins committed:
Remove calls to get_*_dags.

jhtimmins committed:
Add back execution_date

jhtimmins committed:
Add back execution_date

ashb commented on tests/www/test_security.py:
```diff
@@ -29,6 +29,7 @@
 from airflow.models.dag import DAG
 from airflow.security import permissions
 from airflow.www import app as application
+from airflow.www.fab_security.manager import AnonymousUser
```
This is the only change in this file -- is it needed?

ashb commented on airflow/www/fab_security/sqla/manager.py:
```diff
@@ -339,7 +333,7 @@ def filter_roles_by_perm_with_action(self, action_name: str, role_ids: List[int]
         ).all()
 
-    def get_role_permissions_from_db(self, role_id: int) -> List[Permission]:
+    def get_permissions_from_roles(self, role_id: int) -> List[Permission]:
```
```suggestion
    def get_permissions_from_role(self, role_id: int) -> List[Permission]:
```
Should be singular given it takes a single role_id. (Possibly also called `_from_role_id`, but don't mind on that one)

jhtimmins replied:
Its actually not used anymore so should be removed entirely

ashb commented on airflow/www/fab_security/manager.py:
```diff
@@ -85,6 +85,33 @@ def _oauth_tokengetter(token=None):
+        self._roles = roles
+        self._perms = set()
+
+    @property
+    def perms(self):
+        if not self._perms:
+            self._perms = set()
+            for role in self.roles:
+                self._perms.update({(perm.action.name, perm.resource.name) for perm in role.permissions})
+        return self._perms
```
`from airflow.compat.functools import cached_property` and then:

```suggestion
    @cached_property
    def perms(self):
        _perms = set()
        for role in self.roles:
            _perms.update({(perm.action.name, perm.resource.name) for perm in role.permissions})
        return _perms
```
And then `del self.perms` in the `@roles.setter` to clear the cache.

(This is all functionally the same, I just find the delcarative `@cached_property` clearer to read.)

jhtimmins replied:
I'll make this change in a follow up PR

ashb commented on airflow/api_connexion/parameters.py:
```diff
@@ -100,7 +100,7 @@ def apply_sorting(query, order_by, to_replace=None, allowed_attrs=None):
         lstriped_orderby = to_replace.get(lstriped_orderby, lstriped_orderby)
     if order_by[0] == "-":
-        order_by = f"{lstriped_orderby} desc"
+        order_by = desc(text(lstriped_orderby))
```
Unrelated change in this file?

ashb reviewed (commented):
LGTM -- a few small comments to address or say "naaah" to

jhtimmins committed:
Add back comma to match rest of file.

jhtimmins committed:
Add back comma to match rest of file.

jhtimmins committed:
Remove unused permission functions.

jhtimmins committed:
Remove unused permission functions.

jhtimmins committed:
Fix failing tests.

jhtimmins committed:
Fix failing tests.

jhtimmins committed:
Pass user object to current_app.appbuilder.sm.has_all_dags_access.

jhtimmins committed:
Pass user object to current_app.appbuilder.sm.has_all_dags_access.

jhtimmins committed:
Remove attempts to fix query.

jhtimmins committed:
Remove attempts to fix query.

jhtimmins committed:
Update the api_connexion query builders.

jhtimmins committed:
Update the api_connexion query builders.

jhtimmins committed:
Add typing.

jhtimmins committed:
Add typing.

jhtimmins committed:
Apply sorts directly to model objects.

jhtimmins committed:
Apply sorts directly to model objects.

jhtimmins committed:
Apply sorts directly to model objects.

jhtimmins committed:
Apply sorts directly to model objects.

jhtimmins committed:
Standardize custom sort code.

jhtimmins committed:
Standardize custom sort code.

ashb commented on airflow/api_connexion/endpoints/role_and_permission_endpoint.py:
```diff
@@ -68,11 +68,19 @@ def get_roles(*, order_by: str = "name", limit: int, offset: Optional[int] = Non
     appbuilder = current_app.appbuilder
     session = appbuilder.get_session
     total_entries = session.query(func.count(Role.id)).scalar()
+    direction = "desc" if order_by.startswith("-") else "asc"
```
Rather than the double `getattr` we should use the standalone asc/desc functions https://docs.sqlalchemy.org/en/13/core/sqlelement.html?highlight=desc#sqlalchemy.sql.expression.desc

```python
from sqlalchemy import desc, asc
```

and then

```suggestion
    direction = desc if order_by.startswith("-") else asc
```

etc

ashb commented on airflow/api_connexion/endpoints/role_and_permission_endpoint.py:
```diff
@@ -68,11 +68,19 @@ def get_roles(*, order_by: str = "name", limit: int, offset: Optional[int] = Non
     query = session.query(Role)
-    query = apply_sorting(query, order_by, to_replace, allowed_filter_attrs)
-    roles = query.offset(offset).limit(limit).all()
+    roles = query.order_by(getattr(getattr(Role, order_param), direction)()).offset(offset).limit(limit).all()
```
```suggestion
    roles = query.order_by(direction(getattr(Role, order_param)).offset(offset).limit(limit).all()
```

ashb commented on airflow/www/security.py:
```diff
@@ -397,62 +354,35 @@ def has_access(self, action_name, resource_name, user=None) -> bool:
-            self._has_role(['Admin', 'Viewer', 'Op', 'User'])
-            or self._has_perm(permissions.ACTION_CAN_READ, permissions.RESOURCE_DAG)
-            or self._has_perm(permissions.ACTION_CAN_EDIT, permissions.RESOURCE_DAG)
+            self._has_role(['Admin', 'Viewer', 'Op', 'User'], user)
```
(Unrelated to this PR, we should remove this Role check -- those roles _have_ these permissions and by having this here it means you _can't_ remove access to all dags from the stock roles.)

ashb reviewed (changes requested):
Approach looks good and tidies up the SecurityManager code a lot.

One small improvement to make.

ashb committed:
Code review

ashb reviewed (approved):
Made the changes from the review, and have rebased.

lewismc committed:
Augment xcom docs (#20755)

ashb committed:
Fix relationship join bug in FAB/SecurityManager with SQLA 1.4 (#21296)

This is fixed in SQLA 1.4.19, but the fix makes the intent clearer here
anyway.

luciakasman committed:
Docs: Fix task order in overview example (#21282)

fj-sanchez committed:
Update stat_name_handler documentation (#21298)

Previously stat_name_handler was under the scheduler section of the
configuration but it was moved to the metrics section since 2.0.0.

mik-laj committed:
Update recipe for Google Cloud SDK (#21268)

jhtimmins committed:
Fix merge conflicts.

jhtimmins committed:
Make sure joined fields prefetch.

jhtimmins committed:
Dont use cached_property, since its only on > 3.8.
